File: week01/red-sprout/solution04.py
m, n = 0, 0
board = []


def init(_m, _n, sArr):
    global m, n, board

    m = _m
    n = _n
    board = [[''] * n for _ in range(m)]

    for i in range(m):
        for j in range(n):
            board[i][j] = sArr[i][j]


def erase():
    visited = [[False] * n for _ in range(m)]

    for i in range(m - 1):
        for j in range(n - 1):
            c = board[i][j]

            if c == '.':
                continue

            if (c == board[i + 1][j]
                and c == board[i][j + 1]
                and c == board[i + 1][j + 1]):
                visited[i][j] = True
                visited[i + 1][j] = True
                visited[i][j + 1] = True
                visited[i + 1][j + 1] = True

    cnt = 0

    for i in range(m):
        for j in range(n):
            if visited[i][j]:
                cnt += 1
                board[i][j] = '.'

    return cnt


def fall():
    queues = [[] for _ in range(n)]

    for i in range(m):
        for j in range(n):
            if board[i][j] != '.':
                queues[j].append(board[i][j])

    for i in range(m - 1, -1, -1):
        for j in range(n):
            if queues[j]:
                board[i][j] = queues[j].pop()
            else:
                board[i][j] = '.'


def solution(m, n, board):
    init(m, n, board)

    answer = 0

    while True:
        cnt = erase()

        if cnt == 0:
            break

        answer += cnt
        fall()

    return answer

File: week01/red-sprout/solution04.java
import java.util.*;

class Solution {
    int m, n;
    char[][] board;

    void init(int m, int n, String[] sArr) {
        this.m = m;
        this.n = n;
        this.board = new char[m][n];

        for(int i = 0; i < m; ++i) {
            for(int j = 0; j < n; ++j) {
                this.board[i][j] = sArr[i].charAt(j);
            }
        }
    }

    int erase() {
        boolean[][] visited = new boolean[m][n];
        for(int i = 0; i < m - 1; ++i) {
            for(int j = 0; j < n - 1; ++j) {
                char c = board[i][j];
                if(c == '.') continue;
                if(c == board[i + 1][j]
                  && c == board[i][j + 1]
                  && c == board[i + 1][j + 1]) {
                    visited[i][j] = true;
                    visited[i + 1][j] = true;
                    visited[i][j + 1] = true;
                    visited[i + 1][j + 1] = true;
                }
            }
        }

        int cnt = 0;
        for(int i = 0; i < m; ++i) {
            for(int j = 0; j < n; ++j) {
                if(visited[i][j]) {
                    ++cnt;
                    board[i][j] = '.';
                }
            }
        }

        return cnt;
    }

    void fall() {
        Queue<Character>[] queues = new Queue[n];
        for(int j = 0; j < n; ++j) {
            queues[j] = new ArrayDeque<>();
        }

        for(int i = m - 1; i >= 0; --i) {
            for(int j = 0; j < n; ++j) {
                if(board[i][j] != '.') queues[j].offer(board[i][j]);
            }
        }

        for(int i = m - 1; i >= 0; --i) {
            for(int j = 0; j < n; ++j) {
                if(!queues[j].isEmpty()) {
                    board[i][j] = queues[j].poll();
                } else {
                    board[i][j] = '.';
                }
            }
        }
    }

    public int solution(int m, int n, String[] board) {
        init(m, n, board);
        int answer = 0;
        while(true) {
            int cnt = erase();
            if(cnt == 0) break;
            answer += cnt;
            fall();
        }
        return answer;
    }
}
